src/sdl/sdl_touch.c

```diff
@@ -0,0 +1,252 @@
+            case ALLEGRO_EVENT_TOUCH_END:   type = ALLEGRO_EVENT_MOUSE_BUTTON_UP;   break;
+            case ALLEGRO_EVENT_TOUCH_MOVE:  type = ALLEGRO_EVENT_MOUSE_AXES;        break;
+         }
+
+         event.mouse.type      = type;
+         event.mouse.timestamp = timestamp;
+         event.mouse.display   = (ALLEGRO_DISPLAY*)disp;
+         event.mouse.x         = (int)x;
+         event.mouse.y         = (int)y;
+         event.mouse.dx        = (int)dx;
+         event.mouse.dy        = (int)dy;
+         event.mouse.dz        = 0;
+         event.mouse.dw        = 0;
+         if (touch_input->touch_input.mouse_emulation_mode != ALLEGRO_MOUSE_EMULATION_5_0_x) {
+            event.mouse.button = 1;
+         }
+         else {
+            event.mouse.button = id;
+         }
+         event.mouse.pressure  = mouse_state.pressure;
+
+         if (touch_input->touch_input.mouse_emulation_mode != ALLEGRO_MOUSE_EMULATION_5_0_x) {
+            al_set_mouse_xy(event.mouse.display, event.mouse.x, event.mouse.y);
+         }
+
+         _al_event_source_emit_event(&touch_input->touch_input.mouse_emulation_es, &event);
+      }
+      _al_event_source_unlock(&touch_input->touch_input.mouse_emulation_es);
+   }
+}
+
+void _al_sdl_touch_input_event(SDL_Event *e)
+{
+   if (!touch_input)
+      return;
+
+   ALLEGRO_EVENT_TYPE type;
+
+   ALLEGRO_DISPLAY *d = NULL;
+   /* Use the first display as event source if we have any displays. */
+   ALLEGRO_SYSTEM *s = al_get_system_driver();
+   if (_al_vector_size(&s->displays) > 0) {
+      void **v = (void **)_al_vector_ref(&s->displays, 0);
+      d = *v;
+   }
+   if (!d) {
+      return;
+   }
+
+   int touchId = e->tfinger.fingerId;
+
+   touch_input->state.touches[touchId].x = e->tfinger.x;
+   touch_input->state.touches[touchId].y = e->tfinger.y;
+   touch_input->state.touches[touchId].dx = e->tfinger.dx;
+   touch_input->state.touches[touchId].dy = e->tfinger.dy;
+
+   if (e->type == SDL_FINGERDOWN) {
+      type = ALLEGRO_EVENT_TOUCH_BEGIN;
+      touch_input->state.touches[touchId].id = touchId;
+      touch_input->state.touches[touchId].primary = (touch_input->touches == 0);
+      touch_input->touches++;
+   }
+   else if (e->type == SDL_FINGERMOTION) {
+      type = ALLEGRO_EVENT_TOUCH_MOVE;
+   }
+   else if (e->type == SDL_FINGERUP) {
+      type = ALLEGRO_EVENT_TOUCH_END;
+      touch_input->touches--;
+      touch_input->state.touches[touchId].id = -1;
+   } else {
+      return;
+   }
+
+   // SDL2 only returns absolute positions of touches on the input device. This means we have to fake them
+   // in order to make them related to the display, which is what Allegro returns in its API.
+   // This is likely to break in lots of cases, but should work well with non-rotated fullscreen or Wayland windows.
+   generate_touch_input_event(type, e->tfinger.timestamp, touchId,
+                              touch_input->state.touches[touchId].x * al_get_display_width(d),
```

Comment:
I guess this means that it's fullscreen only atm (possibly mobile only?)

Reply:
There is now: https://github.com/SDL-mirror/SDL/commit/e4a845faec1803b3e5c92e2f671f7651cfba4238 :D